quartz/components/ExplorerNode.tsx

```diff
@@ -0,0 +1,205 @@
+    } else {
+      const next = file.path[0]
+      file.path = file.path.splice(1)
+      for (const child of this.children) {
+        if (child.name === next) {
+          child.insert(file, this.depth)
+          return
+        }
+      }
+
+      const newChild = new FileNode(next)
+      newChild.insert(file, this.depth)
+      this.children.push(newChild)
+    }
+  }
+
+  // Add new file to tree
+  add(file: Data, splice: number = 0) {
+    this.insert({ file, path: file.filePath!.split("/").splice(splice) }, -1)
+  }
+
+  // Print tree structure (for debugging)
+  print(depth: number = 0) {
+    let folderChar = ""
+    if (!this.file) folderChar = "|"
+    console.log("-".repeat(depth), folderChar, this.name, this.depth)
+    this.children.forEach((e) => e.print(depth + 1))
+  }
+
+  /**
+   * Get folder representation with state of tree.
+   * Intended to only be called on root node before changes to the tree are made
+   * @param collapsed default state of folders (collapsed by default or not)
+   * @returns array containing folder state for tree
+   */
+  getFolderPaths(collapsed: boolean): FolderState[] {
+    const folderPaths: FolderState[] = []
+
+    const traverse = (node: FileNode, currentPath: string) => {
+      if (!node.file) {
+        const folderPath = currentPath + (currentPath ? "/" : "") + node.name
+        if (folderPath !== "") {
+          folderPaths.push({ path: folderPath, collapsed })
+        }
+        node.children.forEach((child) => traverse(child, folderPath))
+      }
+    }
+
+    traverse(this, "")
+
+    return folderPaths
+  }
+
+  // Sort order: folders first, then files. Sort folders and files alphabetically
+  sort() {
+    this.children = this.children.sort((a, b) => {
+      if ((!a.file && !b.file) || (a.file && b.file)) {
+        return a.name.localeCompare(b.name)
+      }
+      if (a.file && !b.file) {
+        return 1
+      } else {
+        return -1
+      }
+    })
+
+    this.children.forEach((e) => e.sort())
+  }
+}
+
+type ExplorerNodeProps = {
+  node: FileNode
+  opts: Options
+  fullPath?: string
+}
+
+export function ExplorerNode({ node, opts, fullPath }: ExplorerNodeProps) {
+  // Get options
+  const folderBehavior = opts.folderClickBehavior
+  const collapseFolders = opts.folderDefaultState === "collapsed"
+
+  // Calculate current folderPath
+  let pathOld = fullPath ? fullPath : ""
+  let folderPath = ""
+  if (node.name !== "") {
+    folderPath = `${pathOld}/${node.name}`
+  }
+
+  // Helpers to disable pointer-events and opacity for all appropriate elements
+  const isInvisInner = node.depth > 1 && collapseFolders
+  const isInvisOuter = node.depth > 0 && collapseFolders
+
+  return (
+    <div class={`${isInvisOuter ? "no-pointer" : ""}`}>
+      {node.file ? (
+        // Single file node
+        <li key={node.file.slug} class="no-pointer">
+          <a
+            href={`/${node.file.slug}`}
```

Comment:
incorrect to go from `/` as this will break for any Quartz hosted on a subpath (e.g. username.github.io/quartz)

Reply:
should I use the `baseUrl` from `cfg` instead and add paths after based on that instead?

Reply:
that would work but let's not be dependent on `baseUrl`. you should do `resolveRelative`: https://github.com/jackyzha0/quartz/blob/v4/quartz/components/Backlinks.tsx#L15

Reply:
done in 2acdf14570a0ba6a2460ed4ca7bd76012057437d